### PR TITLE
Modify CaloLayer1 emulator to properly handle saturation in all cases

### PR DESCRIPTION
**Pull request by nsmith-**
Will make a backport too.
Firmware already at P5
more info: https://indico.cern.ch/event/655486/contributions/2672892/attachments/1498500/2332845/nsmith_layer1Saturation.pdf

### CODE REVIEW

**Comment by nsmith-** [comment]
Oh sweet 20000 do I get a prize?

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
@nsmith- 
hopefully yes.

**Comment by nsmith-** [comment]
@rekovic can you sign this? (also #20001)

**Comment by rekovic** [comment]
+1

File: L1Trigger/L1TCaloLayer1/python/simCaloStage2Layer1Digis_cfi.py [hunk 1]
```diff
@@ -18,5 +18,5 @@
     unpackEcalMask = cms.bool(False),
     unpackHcalMask = cms.bool(False),
     # See UCTLayer1.hh for firmware version
-    firmwareVersion = cms.int32(1),
+    firmwareVersion = cms.int32(3),
```
Comment (davidlange6) [inline L1Trigger/L1TCaloLayer1/python/simCaloStage2Layer1Digis_cfi.py]: hi @nsmith @rekovic - does this change work for all stage 2 data?

File: L1Trigger/L1TCaloLayer1/src/L1TCaloLayer1FetchLUTs.cc [hunk 2]
```diff
@@ -282,9 +300,27 @@ bool L1TCaloLayer1FetchLUTs(const edm::EventSetup& iSetup,
           if(useCalib) calibratedHFInput *= hfSF.at(phiBin*hfScalePhiBins.size()*12+etBin*12+etaBin);
           if(useLSB) calibratedHFInput /= caloLSB;
 
-          value = calibratedHFInput;
-          if(value > 0xFF) {
-            value = 0xFF;
+          if ( fwVersion > 2 ) {
+            uint32_t absCaloEta = abs(caloEta);
```
Comment (davidlange6) [inline L1Trigger/L1TCaloLayer1/src/L1TCaloLayer1FetchLUTs.cc]: please use std::abs

File: L1Trigger/L1TCaloLayer1/src/L1TCaloLayer1FetchLUTs.cc [hunk 3]
```diff
@@ -282,9 +300,27 @@ bool L1TCaloLayer1FetchLUTs(const edm::EventSetup& iSetup,
...
+            if(absCaloEta > 29 && absCaloEta < 40) {
+              // Divide by two (since two duplicate towers are sent)
+              calibratedHFInput /= 2.;
```
Comment (davidlange6) [inline L1Trigger/L1TCaloLayer1/src/L1TCaloLayer1FetchLUTs.cc]: *=0.5 (and similarly below)

File: L1Trigger/L1TCaloLayer1/src/UCTTower.cc [hunk 4]
```diff
@@ -90,24 +90,37 @@ bool UCTTower::process() {
...
-  uint32_t calibratedET = hcalET;
-  if(hfLUT != 0) {
-    uint32_t etaAddress = (region - NRegionsInCard) * NHFEtaInRegion + iEta;
-    const std::array< uint32_t, 256> a = hfLUT->at(etaAddress);
-    calibratedET = a[hcalET] & 0xFF;
+  if ( fwVersion > 2 ) {
+    uint32_t calibratedET = hcalET;
+    if(hfLUT != 0) {
+      uint32_t etaAddress = (region - NRegionsInCard) * NHFEtaInRegion + iEta;
+      const std::array< uint32_t, 256> a = hfLUT->at(etaAddress);
```
Comment (davidlange6) [inline L1Trigger/L1TCaloLayer1/src/UCTTower.cc]: could this be a const &?

File: L1Trigger/L1TCaloLayer1/python/simCaloStage2Layer1Digis_cfi.py (hunk 1 above)
Comment (nsmith-) [inline L1Trigger/L1TCaloLayer1/python/simCaloStage2Layer1Digis_cfi.py]: Yes and no, the change will not break the ability to re-emulate using old data, but the bit-level agreement will be broken with old (bugged firmware) data.
So far, there has in general been no requirement that L1T emulation can reproduce historic data at bit-level automatically.  (the latest firmware and latest emulator should agree of course)

File: L1Trigger/L1TCaloLayer1/src/L1TCaloLayer1FetchLUTs.cc (hunk 3 above)
Comment (nsmith-) [inline L1Trigger/L1TCaloLayer1/src/L1TCaloLayer1FetchLUTs.cc]: Is this a style thing?
`cat > div.cpp`
```cpp
void div(double& x) {
  x /= 2.;
}
```
`gcc -S div.cpp`
```asm
...
	divsd	%xmm0, %xmm1
...
```
`gcc -O2 -S div.cpp`
```asm
...
	mulsd	LCPI0_0(%rip), %xmm0
...
```

**Comment by nsmith-** [comment]
I couldn't run `scram build code-checks` in the release, is it only in IB now?

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

File: L1Trigger/L1TCaloLayer1/python/simCaloStage2Layer1Digis_cfi.py (hunk 1 above)
Comment (davidlange6) [inline L1Trigger/L1TCaloLayer1/python/simCaloStage2Layer1Digis_cfi.py]: hi @nsmith-interesting- remind me how this code gets used in production workflows?

Comment (nsmith-) [inline L1Trigger/L1TCaloLayer1/python/simCaloStage2Layer1Digis_cfi.py]: it is part of the stage2 L1 trigger emulation, not sure what you are asking specifically

**Comment by davidlange6** [comment]
for example - do we not expect to want to emulate the L1 from 2016 in the future (which is still supported data/MC in 93x)?

**Comment by nsmith-** [comment]
Sure, CaloLayer1 could support this with some era switch, but there are some other pieces of the stage 2 trigger that I know do not have any configuration for backwards compatibility, and saw multiple firmware revisions in 2016 alone.  If I had not added `firmwareVersion` and just patched the code we would likely not be having this conversation.

If we want to make full support of previous trigger algorithms, fine, but maybe in another PR?

**Comment by davidlange6** [comment]
indeed, its already suboptimal that its me and not the L1 group realizing this... anyway, adding the era switch here is straightforward, please try to add it to this PR (obviously the general compatibility is something beyond that)

@rekovic

**Comment by rekovic** [comment]
@davidlange6 
PR #20226 provides new 2017 modifier for L1T `stage2L1Trigger_2017`.  Maybe we can merge it soon. 

One wants to use the correct era flag when emulating L1T for 2016 and 2017 differently.

However, when we re-emulate L1T on already taken data, we also make use of firmware 
version changes that are made in the L1T hardware.  This helps as describe true L1T 
more realistically. 

When it comes to MC production, we rely on one version of L1T emulator (i.e. one fw version), usually the latest greatest.

**Comment by davidlange6** [comment]
> On Aug 29, 2017, at 9:01 AM, rekovic <notifications@github.com> wrote:
> 
> @davidlange6
> PR #20226 provides new 2017 modifier for L1T stage2L1Trigger_2017. Maybe we can merge it soon.
> 
> One wants to use the correct era flag when emulating L1T for 2016 and 2017 differently.
> 
> However, when we re-emulate L1T on already taken data, we also make use of firmware
> version changes that are made in the L1T hardware. This helps as describe true L1T
> more realistically.
> 
> When it comes to MC production, we rely on one version of L1T emulator (i.e. one fw version), usually the latest greatest.
> 

Right - and thats not the correct approach once you have data from the past that you care about.. (as you'd like to simulate it) - of course one has to define how many data taking periods are of interest - but presumably you should maintain one per year (I think 2015 can be forgotten)

**Comment by rekovic** [comment]
@davidlange6 
The switch b/w Run2_2016 and Run2_2017 should be enough granularity for the MC productions.
I am not aware that when we do a "realistic MC Production" where we try to simulate run by run.
Or am I missing something.

**Comment by davidlange6** [comment]
Right, thats what I said...maintain one per year.

**Comment by rekovic** [comment]
@davidlange6  Ok.  

@nsmith- 
So, we should then adjust this PR to take advantage of  `stage2L1Trigger_2016 and 2017`, latter available once PR #20226  is merged.

**Comment by davidlange6** [comment]
hi @rekovic - what is the plan for this PR?

**Comment by nsmith-** [comment]
@davidlange6 it's pending on a change by me, that slipped of the todo list.  I'll do it now.

**Comment by nsmith-** [comment]
Why the heck is 930 not an ancestor of 940?  Ugh, cherry-picking...

**Comment by kpedro88** [comment]
9X0 is never an ancestor of 9(X+1)0

**Comment by nsmith-** [comment]
Well https://github.com/cms-sw/cmssw/releases/CMSSW_9_4_0_pre1 was, but I guess the original base of this PR wasn't 930.  Sorry for the noise

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Comment by davidlange6** [comment]
as for "Why the heck is 930 not an ancestor of 940" - its because the development for 940 starts before 930 is cut (the break point is when the features of 930 are finalized)

**Pull request merged**